Add Kotlin specific assertion helpers

JLLeitschuh opened the pull request:
## Overview

This adds several Kotlin friendly assertions to the standard JUnit assertion API.

Closes #924

## What does this actually add?

The `Assertions.kt` file you see has top-level functions. When the compiler runs it actually creates a "hidden" class called `AssertionsKt` that contains these static methods. This name could be changed with an annotation to something like `KotlinAssertions`. This name will never actually show up in anyone's code but will end up in the compiled bytecode so changing the name of this file will be a breaking change (like changing the name of any class in Java).

## Documentation

Unfortunately, I was unable to get (Dokka)[https://github.com/Kotlin/dokka] to behave correctly so that it would generate JavaDocs. The best that we can do on this front is document these methods in the user guide and show how they can be imported.

---

I hereby agree to the terms of the JUnit Contributor License Agreement.

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests)
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](http://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/meta/API.html)
- [x] Change is documented in the [User Guide](http://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](http://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

JLLeitschuh commented on build.gradle:
```diff
@@ -434,6 +439,13 @@ subprojects { subproj ->
 		}
+
+		kotlin {
+			ktlint("0.9.0")
```
[ktlint](https://ktlint.github.io/) is the only kotlin linter that exists at present. Its pretty well fleshed out.

If you want to have it use tabs instead of spaces for indentation it can be configured using an `.editorconfig` file in the root directory of the project.

JLLeitschuh commented:
Somethings up with JKD9 that I'll try to solve tomorrow.

jbduncan commented on junit-jupiter-api/src/main/kotlin/org/junit/jupiter/api/Assertions.kt:
````diff
@@ -0,0 +1,68 @@
+ * Example usage:
+ * ```kotlin
+ * val exception = assertThrows<IllegalArgumentException> {
+ *     throw IllegalArgumentException("Talk to a 🦆")
````
@JLLeitschuh On my phone, this particular sentence renders as

```kotlin
    throw IllegalArgumentException("Talk to the <box-shape>")
```

where "<box-shape>" renders as a box-shaped character.

Is that what was intended?

JLLeitschuh replied:
Its an emoji 🦆 (duck) character.
https://en.wikipedia.org/wiki/Rubber_duck_debugging

jbduncan replied:
Okay, thanks @JLLeitschuh. 

It renders correctly as a duck emoji on my copy of Chrome, but unfortunately it still renders as a box (the character an application displays when it cannot identify what the character is supposed to be) in Gmail and on my Android phone.

I wonder if it would be more maintainable to swap the duck emoji out for a non-emoji character or phrase.

jbduncan replied:
Sorry, I meant to say that the duck emoji :duck: renders correctly in your comment at https://github.com/junit-team/junit5/pull/1001#discussion_r131382622, but it fails to render as a duck emoji in the GitHub code view on Chrome for me.

This is what it looks like for me in the GitHub code view:
![temp2](https://user-images.githubusercontent.com/2482228/28976860-84135f06-7937-11e7-9ca6-7b5c664acca7.png)

JLLeitschuh replied:
Only in your iPhone browser right? It shows up fine for me on macOS Chrome.

jbduncan replied:
I actually don't own an Apple device; my phone's an Android.

To clarify, it renders as that box shape for me on Windows 7 Chrome and Android Chrome.

JLLeitschuh replied:
Do you want me to change this to something that will render correctly on an IOS device?

jbduncan replied:
I'd quite like it to be changed so it renders correctly on Windows and Android, if that's alright.

(I'm not sure if you saw my earlier comment, but I don't actually own an iOS device. My devices are Windows and Android. :smile:)

JLLeitschuh replied:
Sure, I can fix that.
Question...
Does the emoji in this code example render correctly for you?
http://junit.org/junit5/docs/current/user-guide/#writing-tests-display-names

jbduncan replied:
Hi @JLLeitschuh, yes it does. Here's a screenshot of what it looks like for me on my Windows 7 machine.

![temp7](https://user-images.githubusercontent.com/2482228/29429459-a964ec1a-8388-11e7-8999-bb8bafa52cd8.png)

JLLeitschuh replied:
Huh... This is what it looks like for me on macOS 10.12.6 with chrome.

<img width="974" alt="screen shot 2017-08-17 at 3 32 45 pm" src="https://user-images.githubusercontent.com/1323708/29430166-5e74c9b8-8361-11e7-8d4c-dfebcc123f0a.png">

Anyways, yes,  I can change it.

JLLeitschuh commented:
I'd like to add a kotlin example to the `adoc`s docs if this is something that the maintainers think think this is the right route to go down.

JLLeitschuh commented:
I'm asking in the kotlin slack channel about the problem with the compiler on JDK 9.

JLLeitschuh commented:
Related issue:
https://youtrack.jetbrains.com/issue/KT-19557

AustinShalit commented:
You should add a helper method for `assertThrows` that also takes a message.

sormuras commented:
The JDK 9 "Should be fixed in Kotlin 1.1.4"

JLLeitschuh commented:
This is actually blocked on:
https://github.com/gradle/kotlin-dsl/issues/454

I'm going to have to revert my changes to changing `junit-jupiter-api.gradle.kts` to a `kts` file and figure out how to do that artifact nonsense with groovy.

JLLeitschuh commented:
Okay, this finally passes the CI. Also, @jbduncan I removed the emoji as you requested.

JLLeitschuh commented:
How would you like to see this documented in the User Guide and the Release Notes? Would it make sense to have a kotlin test as an example that is rendered in the documentation?

jbduncan commented:
@JLLeitschuh

> Also, @jbduncan I removed the emoji as you requested.

Great, many thanks! :+1:

JLLeitschuh commented on junit-jupiter-api/src/main/kotlin/org/junit/jupiter/api/Assertions.kt:
```diff
@@ -0,0 +1,68 @@
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
```
Even though this file is called `Assertions.kt` what really happens is this creates a class called `AssertionsKt`. Thus preventing the naming conflict.
If the team would like to be more specific about the name of the class created we could add something like `@file:JvmName("AssertionsForKotlin")` to the top of the file.

This will require some monkeying with spotless due to this issue:
https://github.com/diffplug/spotless/issues/136

Does the team mind that the kotlin compiler just generates a class and uses its own naming convention? If not then I don't need to worry about the spotless bug.

marcphilipp replied:
Is this naming convention likely to change in the future?

JLLeitschuh replied:
Nope, it shouldn't.

JLLeitschuh commented on junit-jupiter-api/src/main/kotlin/org/junit/jupiter/api/Assertions.kt:
```diff
@@ -0,0 +1,99 @@
+ */
+@API(Experimental)
+inline fun <reified T : Throwable> assertThrows(noinline message: () -> String, noinline executable: () -> Unit): T =
+    Assertions.assertThrows(T::class.java, Executable(executable), Supplier(message))
```
The API that this exposes is inverted to what the Java API exposes. There were two options:

1. Putting the message first.
    Signature:
    ```kotlin
    inline fun <reified T : Throwable> assertThrows(noinline message: () -> String, noinline executable: () -> Unit): T
    ```
    When called:
    ```kotlin
    assertThrows<IllegalArgumentException>({ "Should throw an Exception" }) {
        throw IllegalArgumentException("Talk to a duck")
    }
    ```
2. Putting the message second (matching the Java API)
    Signature:
    ```kotlin
    inline fun <reified T : Throwable> assertThrows(noinline executable: () -> Unit, noinline message: () -> String): T
    ```
    When called
    ```kotlin
    assertThrows<IllegalArgumentException>({
        throw IllegalArgumentException("Talk to a duck")
    }) { "Should throw an Exception" }
    ```

Personally I thought that option 1 was easier to read and was more intuitive.

If you disagree and think I should change it please let me know.

JLLeitschuh commented on junit-jupiter-engine/src/test/java/org/junit/jupiter/api/AssertionsAssertAllTests.java:
```diff
@@ -149,7 +149,7 @@ void assertAllWithExecutableThatThrowsBlacklistedException() {
 
 	@SafeVarargs
-	private static void assertExpectedExceptionTypes(MultipleFailuresError multipleFailuresError,
+	static void assertExpectedExceptionTypes(MultipleFailuresError multipleFailuresError,
```
This method is incredibly useful and I would 100% support making it part of the public API.

JLLeitschuh commented:
> You should add a helper method for assertThrows that also takes a message.
Thanks @AustinShalit this should be resolved now.

JLLeitschuh commented:
The change has now been documented in the UserGuide. Here's the UserGuide PDF that I generated from 1a0bce3aec55740bd667ca74f97e3da362e53116
[UserGuide.pdf](https://github.com/junit-team/junit5/files/1265464/UserGuide.pdf)

Coderay doesn't have syntax highlighting for Kotlin so I used the Java syntax highlighter.
This can be fixed when asciidoctor is switched to using coderay. #777

JLLeitschuh commented on documentation/src/test/java/example/AssertionsDemo.java:
```diff
@@ -143,23 +143,3 @@ private static String greeting() {
 // end::user_guide[]
 // @formatter:on
-
-class Person {
```
I couldn't access this class in `AssertionsDemoKotlin` without making it public. It was just easier to make `Person` a kotlin data class that both tests use.

JLLeitschuh commented:
So, I tried figuring out how to get my new kotlin methods to show up in the generated javadocs and it ended up being much more complicated than I'd expected.

I ended up opening this issue against dokka: https://github.com/Kotlin/dokka/issues/211

JLLeitschuh commented:
Ended up discovering a problem with the JDK 9 javadoc generator's interaction with kotlin bytecode as a byproduct of this PR:
https://youtrack.jetbrains.com/issue/KT-20025

sormuras commented:
A very detailed issue description. Keen to see how the OpenJDK folks react to this.

JLLeitschuh commented:
> A very detailed issue description. Keen to see how the OpenJDK folks react to this.

Thanks, it took me a few hours to figure out the problem and narrow it down to the minimal code that was the cause. I bit the bullet and downloaded JDK 9 on my machine  so I could solve the problem fully.

---

I'm happy with this PR enough to consider it "Done" in my eyes. This PR has all of the functionality I detailed in my original issue. I welcome all feedback on this.
I would be thrilled to see this PR get into 5.0 GA.

sbrannen commented:
Tentatively slated for 5.0 GA in order to allow the team to assess this PR before GA (even though it might get moved back to 5.1).

smoyer64 commented:
In #1009 it was decided that adding assertions to help TestNG users migrate didn't belong in the core project.  Isn't adding Kotlin assertions to org.junit.jupiter.api a contradiction to the idea that the core system is JUnit 5 and dependency-free?

I'm not against either of these proposals but I do think that bloat in JARs generally happens so gradually that you don't notice until it's too late.  Why not make this another module that people can choose to import (or won't import if they're not using Kotlin)?

JLLeitschuh commented:
It's an optional dependency: 
https://github.com/junit-team/junit5/issues/924#issuecomment-318376186
https://github.com/junit-team/junit5/pull/1001/files#diff-4e064bfaa486310969535672524d34bcR4

This way, if you have Kotlin on the classpath you can call these methods. However, if you don't its a small amount of unused bytecode.

JLLeitschuh commented:
Additional documentation information; I ended coming up with a minimal demo that demonstrates the youtrack/oracle bug that I've opened.
That demo can be found here: https://github.com/JLLeitschuh/kotlin-jdk9-javadoc-bug-demo

JLLeitschuh commented:
I've updated this PR to resolve merge conflicts with master. I changed my release notes to be under `5.0.1`. I wasn't sure if you wanted them under that version or `5.1.0`. I can change them if necessary.

Any thoughts on a timeline on this PR?

JLLeitschuh commented on build.gradle:
```diff
@@ -8,7 +8,7 @@ buildscript {
 	}
 	dependencies {
-		classpath 'com.diffplug.spotless:spotless-plugin-gradle:3.5.1'
+		classpath 'com.diffplug.spotless:spotless-plugin-gradle:3.5.2'
```
Needed to bump the version of spotless so that I [could do this](https://github.com/junit-team/junit5/pull/1001/files#diff-70dc7ae1f1905b194a2aaa9505fc4177R10) and not have spotless try to add the header in the wrong place.

See this issue: https://github.com/diffplug/spotless/issues/136
That I resolved here: https://github.com/diffplug/spotless/pull/139

sbrannen commented:
> Any thoughts on a timeline on this PR?

I hope we'll get a chance to look at for 5.1, but we haven't had a team meeting yet to triage the backlog.

JLLeitschuh commented:
Should I be marking the API as since `5.0.1` or as `5.1`? I can do either.

sormuras commented:
Use 5.1 please, as it is an addition to the Jupiter feature set.

Btw. `master` now contains a `release-notes-5.1.0-M1.adoc` since https://github.com/junit-team/junit5/commit/aa8083d1fcb6404509b67aa23ddafb30c166fcab

JLLeitschuh commented:
Status check on this PR.

JLLeitschuh commented:
@marcphilipp @jbduncan @sormuras @sbrannen 

Is there anything besides the current merge conflict preventing this from being merged?
I feel like this PR is stuck in limbo. If you have any other problems with it please let me know so I can fix those up.

jbduncan commented:
@JLLeitschuh Hmm. As far as I'm aware, there's nothing else stopping this PR from being merged. I can only assume that the JUnit 5 team are busy with IRL things and have prioritised other tasks like Java 9 compatibility over this PR.

So I'm sorry to say that I don't know for sure why this PR is in limbo. :/

marcphilipp commented:
> Is there anything besides the current merge conflict preventing this from being merged?
I feel like this PR is stuck in limbo. If you have any other problems with it please let me know so I can fix those up.

Sorry to keep you waiting! The current plan is to get 5.1 M1 out with JDK 9 module scanning/selector support and then reconsider open PRs and the 5.1 backlog. As I've stated in https://github.com/junit-team/junit5/issues/924#issuecomment-331457939, we want to add support to make writing assertions in Kotlin easier. So, please bear with us a little longer. We'll eventually get to this. We do not expect you to keep this branch up-to-date every time we commit something on master.

JLLeitschuh commented:
Thank you for this explanation. Really helpful. I would have loved to see this in `5.1.0-M1` but I can wait if necessary.

I hadn't realized that you had made the team decision explicit in that comment. Thank you for pointing it out.

Please let me know when you want me to update this PR to resolve merge conflicts in preparation for merging.

marcphilipp commented:
Hi @JLLeitschuh, we've finally decided to merge this pull request. However, there are still a few problems that need to be fixed:

1. We had to hardcode `kotlinOptions{ jvmTarget = '1.8' }` which is fine.
2. Spotless fails when this PR is rebased with the current master because on master the indentatio of all Kotlin files was changed to use tabs already. So, please add an `.editorconfig` to make `ktlint` use tabs, too.

JLLeitschuh commented:
You can't make ktlint use tabs, it's only configured to use spaces. The only configuration that there seems to be is to use either two spaces or four.
https://github.com/shyiko/ktlint/issues/43#issuecomment-304953280

I'm going to go with the ktlint default of using four spaces.

JLLeitschuh commented on junit-jupiter-api/src/main/kotlin/org/junit/jupiter/api/Assertions.kt:
```diff
@@ -0,0 +1,93 @@
+/*
```
The compiler will generate a class called `AssertionsKt` to store these methods inside of. Is this acceptable? We can rename the generated class if necessary.

marcphilipp replied:
I think that's fine

JLLeitschuh replied:
Cool!

JLLeitschuh commented on documentation/src/docs/asciidoc/writing-tests.adoc:
```diff
@@ -94,6 +94,16 @@ are `static` methods in the `{Assertions}` class.
+used in https://kotlinlang.org/[Kotlin]. All JUnit Jupiter Kotlin assertions are top-level
+functions in the `org.junit.jupiter.api` package.
+
+// TODO: Change to using kotlin language highlighting after switch to rouge syntax highlighter
```
What is the status of switching to the rouge syntax highlighter?

marcphilipp replied:
Still blocked by https://github.com/asciidoctor/asciidoctor/issues/1040.

JLLeitschuh commented:
This has been updated per your comments. Let me know if you have anything else you need me to resolve.

shyiko commented:
@JLLeitschuh `.editorconfig` below makes ktlint ignore indentation completely (allowing tabs). This is  not recommended as both [Jetbrains & Google advocate 4-space-indentation](https://github.com/shyiko/ktlint/issues/4) but is perfectly valid as far a ktlint is concerned. 
```ini
[*.{kt,kts}]
indent_size=unset
```

JLLeitschuh commented:
@marcphilipp This is your project, it's up to you to decide.

@shyiko It would be good if that were documented somewhere (site or readme). Thanks for jumping in here!

marcphilipp commented on documentation/src/docs/asciidoc/writing-tests.adoc:
```diff
@@ -94,6 +94,16 @@ are `static` methods in the `{Assertions}` class.
 include::{testDir}/example/AssertionsDemo.java[tags=user_guide]
 ----
 
+Junit Jupiter also comes with a few assertion methods that lend themselves well to being
```
`JUnit`

marcphilipp commented on junit-jupiter-api/src/main/kotlin/org/junit/jupiter/api/Assertions.kt:
```diff
@@ -0,0 +1,93 @@
+ * [Stream] of functions to be executed.
+ */
+internal typealias ExecutableStream = Stream<() -> Unit>
+internal fun ExecutableStream.convert() = map { Executable(it) }
```
Can we make these private?

JLLeitschuh replied:
Effectively they are. `internal` is different from java's "package private".
"Package private" in java means that if I (a consumer) create a file in a package called `org.junit.api` then I can access "package private" things. The same is not true in Kotlin for `internal`.
Internal is `internal` only to source code that is compiled together.

I can make them private if you wish though.

JLLeitschuh replied:
So I was able to make them private.
I remember that in earlier versions of kotlin you couldn't call `private` things inside of `inline` functions. I'm not certain if that limitation still exists. But it doesn't seem to be a problem here.

JLLeitschuh commented:
This travis failure doesn't look like it's created by me:
```
A problem occurred evaluating root project 'junit5'.
> Could not resolve all files for configuration 'classpath'.
   > Could not resolve de.schauderhaft.degraph:degraph-check:0.1.4.
     Required by:
         unspecified:unspecified:unspecified
      > Could not resolve de.schauderhaft.degraph:degraph-check:0.1.4.
         > Could not get resource 'https://repo1.maven.org/maven2/de/schauderhaft/degraph/degraph-check/0.1.4/degraph-check-0.1.4.pom'.
            > Could not GET 'https://repo1.maven.org/maven2/de/schauderhaft/degraph/degraph-check/0.1.4/degraph-check-0.1.4.pom'.
               > Host name 'repo1.maven.org' does not match the certificate subject provided by the peer (CN=repo.maven.apache.org, O="Sonatype, Inc", L=Fulton, ST=MD, C=US)
```

sormuras commented:
Restarting didn't solve the issue -- seems related to https://issues.sonatype.org/browse/MVNCENTRAL-1369

sormuras commented:
Same setup using Gradle 4.3.1 on master does work: https://travis-ci.org/junit-team/junit5/builds/300986101#L643

I don't see any changes in this PR that might effect the SSL configuration. `https://repo1.maven.org` responded as expected. Curious whether the next build will also fail with the same issue...

sormuras commented:
Same on master now... _sigh_

mkobit commented:
https://issues.sonatype.org/browse/MVNCENTRAL-2870 is the issue on the Maven Central side, https://github.com/gradle/gradle/issues/3463 is for Gradle to use the proper CDN URL.

marcphilipp commented:
Merged in 75c3d17. Thanks for your perseverance, @JLLeitschuh! 👍

JLLeitschuh commented:
Wooot!!!! This makes me so happy!!!
